jerry-core/ecma/operations/ecma-objects-general.c

```diff
@@ -658,71 +658,73 @@ ecma_op_general_object_define_own_property (ecma_object_t *obj_p, /**< the objec
   ecma_property_types_t current_type = ECMA_PROPERTY_GET_TYPE (current_p);
   const bool is_current_data_descriptor = (current_type == ECMA_PROPERTY_TYPE_NAMEDDATA);
   const bool is_current_accessor_descriptor = (current_type == ECMA_PROPERTY_TYPE_NAMEDACCESSOR);
+  const bool current_property_configurable = ecma_is_property_configurable (current_p);
 
   JERRY_ASSERT (is_current_data_descriptor || is_current_accessor_descriptor);
 
-  bool is_every_field_in_desc_also_occurs_in_current_desc_with_same_value = true;
-  if (property_desc_p->is_value_defined)
+  if (!current_property_configurable)
   {
-    if (!is_current_data_descriptor
-        || !ecma_op_same_value (property_desc_p->value,
-                                ecma_get_named_data_property_value (current_p)))
+    bool is_every_field_in_desc_also_occurs_in_current_desc_with_same_value = true;
+
+    if (property_desc_p->is_value_defined)
     {
-      is_every_field_in_desc_also_occurs_in_current_desc_with_same_value = false;
+      if (!is_current_data_descriptor
+          || !ecma_op_same_value (property_desc_p->value,
+                                  ecma_get_named_data_property_value (current_p)))
+      {
+        is_every_field_in_desc_also_occurs_in_current_desc_with_same_value = false;
+      }
     }
-  }
 
-  if (property_desc_p->is_writable_defined)
-  {
-    if (!is_current_data_descriptor
-        || property_desc_p->is_writable != ecma_is_property_writable (current_p))
+    if (property_desc_p->is_writable_defined)
     {
-      is_every_field_in_desc_also_occurs_in_current_desc_with_same_value = false;
+      if (!is_current_data_descriptor
+          || property_desc_p->is_writable != ecma_is_property_writable (current_p))
+      {
+        is_every_field_in_desc_also_occurs_in_current_desc_with_same_value = false;
+      }
```

Comment:
These `if (a) { if (!b || !c) {} }` constructs are a very interesting way of writing `if (a && (!b || !c))`. (Used all over here.)